File: bc-tabulate.py
import argparse
import os.path
import sys

parser = argparse.ArgumentParser(description='Tabulate barcode counts across samples')
parser.add_argument('samplefiles', metavar='SAMPLE.TXT', nargs='+', help='Input barcode count (tab-delimited, barcode<TAB>count)')
parser.add_argument('-o', '--output', required=True, help='Output filename for tab-delimited count table')
parser.add_argument('-m', '--mintotal', type=int, help='Minimum total read count across all samples to report barcode')
parser.add_argument('-n', '--minsamples', type=int, help='Minimum number of samples with barcode to report')
parser.add_argument('--mininsample', type=int, help='Minimum read count in at least one sample to report barcode')
parser.add_argument('--omitfile', help='Output filename for omitted barcodes')
args = parser.parse_args()

(root, ext) = os.path.splitext(args.output)
omitfile = '{}-omitted{}'.format(root, ext) if args.omitfile is None else args.omitfile

with open(args.output, 'w') as reportout:
    with open(omitfile, 'w') as omitout:
        barcode_sample_counts = {}
        samples = []
        for samplefile in args.samplefiles:
            sample = os.path.basename(samplefile)
            samples.append(sample)
            for l in open(samplefile, 'r'):
                (barcode, countstr) = l.strip().split('\t')
                count = int(countstr)
                barcode_samples = barcode_sample_counts.get(barcode, {})
                barcode_samples[sample] = count
                barcode_sample_counts[barcode] = barcode_samples

        for out in [reportout, omitout]:
            out.write('barcode')
            for sample in samples:
                out.write('\t{}'.format(sample))
            out.write('\n')
                
        for (barcode, sample_counts) in barcode_sample_counts.items():
            keep_total = ( (args.mintotal is None) or (sum(sample_counts.value() >= args.mintotal)) )
            keep_insample = ( (args.minsamples is None) or (len(sample_counts) >= args.minsamples) )
            keep_minin = ( (args.mininsample is None) or (max(sample_counts.values()) >= args.mininsample) )
            keep = keep_total and keep_insample and keep_minin
            out = reportout if keep else omitout
            out.write('{}'.format(barcode))
            for sample in samples:
                out.write('\t{}'.format(sample_counts.get(sample, 0)))
            out.write('\n')

File: barcode-assign-8b38f7a/src/bc_tabulate.rs
extern crate barcode_assign;
extern crate clap;

use std::io::Write;

use clap::{App, Arg};

use barcode_assign::bc_tabulate::*;

fn main() {
    let matches = App::new("bc-tabulate")
        .version("1.0")
        .author("Nick Ingolia <ingolia@berkeley.edu>")
        .about("Tabulate barcode counts from many different samples")
        .arg(
            Arg::with_name("inputs")
                .index(1)
                .multiple(true)
                .takes_value(true)
                .required(true),
        )
        .arg(
            Arg::with_name("output")
                .short("o")
                .long("output")
                .value_name("OUTPUT.TXT")
                .help("Output filename for tab-delimited count table")
                .takes_value(true)
                .required(true),
        )
        .arg(
            Arg::with_name("mintotal")
                .short("m")
                .long("mintotal")
                .value_name("#READS")
                .help("Minimum total read count across all samples to report barcode")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("minsamples")
                .short("n")
                .long("minsamples")
                .value_name("#SAMPLES")
                .help("Minimum number of samples with barcode to report")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("mininsample")
                .long("mininsample")
                .value_name("#READS")
                .help("Minimum read count in at least one sample to report barcode")
                .takes_value(true),
        )
        .arg(
            Arg::with_name("omitfile")
                .long("omitfile")
                .value_name("OMITTED.TXT")
                .help("Output filename for omitted barcodes")
                .takes_value(true),
        )
        .get_matches();

    fn parse_int(value: &str) -> usize {
        value
            .parse::<usize>()
            .unwrap_or_else(|e| panic!("Bad whole number argument {:?}: {}", value, e))
    }

    let cli = CLI {
        inputs: matches
            .values_of("inputs")
            .unwrap()
            .map(String::from)
            .collect(),
        output: matches.value_of("output").unwrap().to_string(),
        mintotal: matches.value_of("mintotal").map(parse_int),
        minsamples: matches.value_of("minsamples").map(parse_int),
        mininsample: matches.value_of("mininsample").map(parse_int),
        omitfile: matches.value_of("omitfile").map(String::from),
    };

    match cli.run() {
        Ok(_) => (),
        Err(e) => {
            std::io::stderr()
                .write(format!("{}\n", e).as_bytes())
                .unwrap();
            std::process::exit(1);
        }
    }
}
